### ISSUE TITLE
ZNC disconnects every 33 minutes from freenode server

### ISSUE THREAD

**Issue by damfino** [znc/znc#487]
I'm running version 1.2-6 of ZNC on DSM 4.3.
If a client is attached to the bouncer everything works just fine. As soon as I close my client, ZNC starts disconnecting from the freenode server every 33 minutes or so. Like this:
[21:51:28] Disconnected from IRC (kornbluth.freenode.net +7070)
[21:51:43] Connected to IRC (kornbluth.freenode.net +7070)
[22:25:29] Disconnected from IRC (kornbluth.freenode.net +7070)
[22:25:45] Connected to IRC (kornbluth.freenode.net +7070)
And so on.
I've already asked for help on SynoCommunity but they told me that "this is most likely not package related I suggest you seek help upstream to ZNC directly."
Is there some settings I should check? Thanks.

**Comment by alolis** [comment]
Is this resolved? I still see the same behavior on Freenode with my ZNC

**Comment by Mikaela** [comment]
@alolis `/msg *status version`

**Comment by alolis** [comment]
@Mikaela 

````
 <*status>	ZNC 1.2 - http://znc.in
 <*status>	IPv6: yes, SSL: yes, DNS: threads
````

**Comment by Mikaela** [comment]
@alolis ZNC 1.2 is 1.5 years old and not supported by ZNC. Please upgrade to 1.6.0 or 1.6.1rc.

**Comment by alolis** [comment]
@Mikaela , I will give it a go, thanks

**Comment by Mikaela** [comment]
@alolis did newer version fix the issue?

I am wondering if this could be closed.

**Comment by alolis** [comment]
@Mikaela , well, I have been using the following version:

`<*status>	ZNC - 1.6.0 - http://znc.in`

and indeed the situation is better. I do notice disconnects every now and then like the ones I was noticing before but definitely not as often. I am using znc on Freenode, not sure if the irc server plays a part on the problem.

A friend has mentioned the exact same problem on Freenode as well. If you need a specific test I can try it out.

**Comment by Mikaela** [comment]
I think this might be closable then depending on how often that is. Freenode often has DDoS issues and the previous one was few week(s) ago, I think.

**Comment by alolis** [comment]
Maybe you are right. I haven't seen anything weird the last 3 days. You can close the issue if you want and I will just open a new one if i notice something out of the ordinary.

**Comment by jpnurmi** [comment]
Please leave a comment if the problem persists.

**Comment by ransagy** [comment]
Apologies in advance for resurrecting such an old issue, But i seem to have come across the very same thing, on multiple occasions.

At first, I was running the znc-msvc builds, namely a ZNC 1.5 build running on a Windows 10 Intel NUC machine. Obviously that is quite old and somewhat unsupported, So i ended up running a self-compiled 1.6.3 from source, running in my local LAN on Raspbian (Raspberry Pi 3). The former Windows machine was on WiFi, the Pi is directly attached by ethernet to the router.

Machine(s) and/or connection doesn't seem to be dropping when this happens.
As noted, a client attached seems to alleviate this in some way, possibly because of a client's own pings to ZNC.

In both cases, I used the AdiIRC client for Windows, But I've also tried others, and as said, the problem is far less frequent when a client is attached.

Version output is as follows:
[22:38:56]  *status > ZNC 1.6.3 - http://znc.in
[22:38:56]  *status > IPv6: yes, SSL: yes, DNS: threads, charset: yes

The relevant excerpt from znc -D :

[2016-05-30 22:02:15.926477] (chaosblade/freenode) IRC -> ZNC [PING :sinisalo.freenode.net]
[2016-05-30 22:02:15.926710] (chaosblade/freenode) ZNC -> IRC [PONG :sinisalo.freenode.net]
[2016-05-30 22:04:39.926008] (chaosblade/freenode) IRC -> ZNC [PING :sinisalo.freenode.net]
[2016-05-30 22:04:39.926238] (chaosblade/freenode) ZNC -> IRC [PONG :sinisalo.freenode.net]
[2016-05-30 22:09:29.370408] (chaosblade/freenode) ZNC -> IRC [PING :ZNC]
[2016-05-30 22:09:29.469420] (chaosblade/freenode) ZNC -> IRC [QUIT :ZNC 1.6.3 - http://znc.in]

When IRC asks ZNC for pings, it seems to be fine. When ZNC asks for one, something goes awry.
This could possibly be my LAN or some other issue locally related, but i don't notice any other issues with networking at the same time; I'm also on QuakeNet in this case and it does not disconnect, Nor do any of my other non-IRC services experience this.

Any help in looking into this will be greatly appreciated.

**Comment by psychon** [comment]
All I can see from that log is: ZNC tries to PING the IRC server. When writing that to the socket, the kernel returns "connection reset by peer" (so neither of the PING nor the QUIT are actually sent). That means that the connection is already dead at this point and just because it writes something does ZNC (and the kernel) now find out about this.

Technical details:
So I guess the following happens: ZNC sends the PING. Kernel says "ok" and sends a TCP packet to the other end of the connection with this data. The other end replies with a a packet with the RST flag set ("reset"), because it does not know about this TCP connection. Kernel informs ZNC that the "connection was reset by peer" (packet with RST flag set).
On why the connection is dead: No idea. Something on the other end of the connection or in the network (lots of packet loss). Neither seems plausible to me.

**Comment by psychon** [comment]
Some other ideas:
- Since I doubt that the computer that the IRC server runs on crashed: What message did others on IRC see you quit with? Ping timeout? That would mean that the IRC server sent a PING which never reached ZNC.
- What's your normal round-trip-time to the IRC server (ping)? In the above log, there is about 0.1 seconds between sending the PING and noticing that the connection is dead. If this were much lower, then it could have been some local NAT router interfering.

**Comment by ransagy** [comment]
Thanks for the quick response and info.

It doesn't seem like the ZNC process or server crash, but I've switched ZNC
to run in -D for a day or so, writing it's log to a file. Might get more
useful information from there.

My quit messages were indeed ping timeouts, every single time. On average,
freenode disconnects every hour or so, while quakenet happens (maybe) once
a day or two.

From the quick look i had, the round trips seem similar, around 0.1s. I'll
have a look at the extended debug log later today and see.

Any suggestions on thing i could try to check to pinpoint the problem
further?

**Comment by psychon** [comment]
I think that means that the problem didn't occur in your computer. There must have been something on the network. To ZNC it looks like Freenode stopped sending PINGs, to Freenode it looks like ZNC stopped answering PINGs.

**Comment by ransagy** [comment]
Here's a more through log, grep'd to contain only the relevant parts.

[znc_pingpong_sockerror.txt](https://github.com/znc/znc/files/296434/znc_pingpong_sockerror.txt)

The SockError 101's are due to not resolving the hostname during the reconnection, i believe.
I wonder if there's something i can do about it from ZNC's side or is it something entirely up to freenode.

**Comment by psychon** [comment]
I don't see reconnects in that log, but I guess the following:
```
E:\znc.log:2604:[2016-06-01 09:58:00.366238] (chaosblade/freenode) ZNC -> IRC [PING :ZNC]
E:\znc.log:2605:[2016-06-01 09:58:00.433719] IRC::chaosblade::freenode == SockError(104 Connection reset by peer)
E:\znc.log:2613:[2016-06-01 09:58:00.629373] IRC::chaosblade::freenode == SockError(101 Network is unreachable)
```
means that the connection first broke and then the reconnect failed because "Network is unreachable". That points a lot towards network issues. Since apparently only you see this, the network issues likely are between the computer you run ZNC on and "the internet".

(This pattern is visible several times)

Still, nothing ZNC-related.

**Comment by ransagy** [comment]
Those seemed related to the IRC server trying to find my hostname. It
succeeds mostly.
Either way, Looks like i'll have to take it up with freenode or try to lock
into some specific servers of theirs.

Thanks again for the help.


Regards,
Ran Sagy
Software Engineer

**Comment by leonklingele** [comment]
I can confirm that this has been fixed in latest master (e86efb08bd606ea7b38749904e2347237dede6e3).
However the latest release — 1.6.5 — is still broken.
